pydoctor/astbuilder.py

```diff
@@ -426,6 +462,51 @@ def _handleOldSchoolMethodDecoration(self, target: str, expr: Optional[ast.expr]
                     target_obj.kind = model.DocumentableKind.CLASS_METHOD
                 return True
         return False
+    
+    def _warnsConstantAssigmentOverride(self, obj: model.Attribute, lineno_offset: int) -> None:
+        obj.report(f'Assignment to constant "{obj.name}" overrides previous assignment '
+                    f'at line {obj.linenumber}, the original value will not be part of the docs.', 
+                            section='ast', lineno_offset=lineno_offset)
+                            
+    def _warnsConstantReAssigmentInInstance(self, obj: model.Attribute, lineno_offset: int = 0) -> None:
+        obj.report(f'Assignment to constant "{obj.name}" inside an instance is ignored, this value will not be part of the docs.', 
+                        section='ast', lineno_offset=lineno_offset)
+
+    def _handleConstant(self, obj: model.Attribute, value: Optional[ast.expr], lineno: int) -> None:
+        """Must be called after obj.setLineNumber() to have the right line number in the warning."""
+        
+        if is_attribute_overridden(obj, value):
+            
+            if obj.kind in (model.DocumentableKind.CONSTANT, 
+                                model.DocumentableKind.VARIABLE, 
+                                model.DocumentableKind.CLASS_VARIABLE):
+                # Module/Class level warning, regular override.
+                self._warnsConstantAssigmentOverride(obj=obj, lineno_offset=lineno-obj.linenumber)
+            else:
+                # Instance level warning caught at the time of the constant detection.
+                self._warnsConstantReAssigmentInInstance(obj)
+
+        obj.value = value
+        
+        obj.kind = model.DocumentableKind.CONSTANT
+
+        # A hack to to display variables annotated with Final with the real type instead.
+        if is_using_typing_final(obj):
+            if isinstance(obj.annotation, ast.Subscript):
+                # Will not display as "Final[str]" but rather only "str"
+                ann_slice = obj.annotation.slice
+                if isinstance(ann_slice, ast.Index):
+                    obj.annotation = ann_slice.value
+                else:
+                    # Python 3.9
+                    obj.annotation = ann_slice
```

Comment:
In Python 3.8's documentation of the `ast` module, it is stated that a subscript's slice can be `Slice`, `ExtSlice` or `Index`. Therefore it is not valid to conclude that if `ann_slice` is not an `Index`, we are running under Python 3.9+.

Of course in practice people shouldn't be using for example `Final[str, int:3:2]` as an annotation, but if they do for some reason, pydoctor should handle it gracefully.

I think that for `Slice` and `ExtSlice`, we can issue a warning that the annotation is invalid. And then proceed by not using `ann_slice`.

If `ann_slice` is neither `Index`, `Slice` nor `ExtSlice`, then we do know we are on Python 3.9+ and if mypy doesn't consider `ann_slice` an expression then, we can `assert` that it is.

Reply:
I think we're good on this issue